README.md
# babel-preset

[![NPM version][npm-image]][npm-url] [![Downloads][downloads-image]][npm-url] [![Build Status][build-status-image]][build-status-url] [![Coverage Status][codecov-image]][codecov-url] [![Dependency status][david-dm-image]][david-dm-url] [![Dev Dependency status][david-dm-dev-image]][david-dm-dev-url]

[npm-url]:https://npmjs.org/package/@moxy/babel-preset
[npm-image]:https://img.shields.io/npm/v/@moxy/babel-preset.svg
[downloads-image]:https://img.shields.io/npm/dm/@moxy/babel-preset.svg
[build-status-url]:https://github.com/moxystudio/babel-preset/actions
[build-status-image]:https://img.shields.io/github/workflow/status/moxystudio/babel-preset/Node%20CI/master
[codecov-url]:https://codecov.io/gh/moxystudio/babel-preset
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/babel-preset/master.svg
[david-dm-url]:https://david-dm.org/moxystudio/babel-preset
[david-dm-image]:https://img.shields.io/david/moxystudio/babel-preset.svg
[david-dm-dev-url]:https://david-dm.org/moxystudio/babel-preset?type=dev
[david-dm-dev-image]:https://img.shields.io/david/dev/moxystudio/babel-preset.svg

[Babel](https://babeljs.io/) preset to be used at MOXY.

## Installation

```ssh
$ npm install @moxy/babel-preset @babel/core --save-dev
```

If you are using Jest for testing, you also need to install [`babel-jest`](https://github.com/facebook/jest/tree/master/packages/babel-jest):

```ssh
$ npm install babel-jest --save-dev
```

## Motivation

Projects developed at MOXY often use new JavaScript language features that may not be supported in the targets they will run. This preset provides a shareable Babel config that:

- Allows you to use the latest JavaScript features and transpile only what is not already supported by your targets, thanks to [`preset-env`](https://www.npmjs.com/package/babel-preset-env)
- Enables [`class-properties`](https://www.npmjs.com/package/@babel/plugin-proposal-class-properties)
- Optionally enables React, transforming JSX to standard JavaScript
- Uses [`add-module-exports`](https://www.npmjs.com/package/babel-plugin-add-module-exports) to get around [babel#2212](https://github.com/babel/babel/issues/2212)
- Enables [`babel-plugin-lodash`](https://www.npmjs.com/package/babel-plugin-lodash)


## Do I need to transpile?

There has been [discussion](https://github.com/parcel-bundler/parcel/pull/559#discussion_r161926651) in the community about libraries not being compiled, leaving all compilation efforts to top-level projects consuming them. This makes sense, since developers know what platforms their top-level project target and are able to compile their dependencies accordingly. Furthermore, library maintainers are constantly having to update their compilation options as new features are accepted into different stages of the specification, which creates significant and unnecessary overhead.

Problems arise, however, in libraries which target both Node.js and browser, or if non-standard JavaScript is being used, such as [proposals](https://github.com/tc39/proposals) or [JSX](https://reactjs.org/docs/introducing-jsx.html). In those situations, library authors are required to transpile their libraries code to offer `CommonJS` and `ES` module variants or to transform non-standard JavaScript to standard JavaScript.

**In conclusion**:

1. For libraries, you need to transpile if you want to publish both in CommonJS and ES or if there are non-standard JavaScript language features being used
2. For top-level projects, you need to transpile both your code and your dependencies if the JavaScript language features being used are not supported by your targets

## Usage

### 1. Choose a preset-type

There're two preset types available for you to use:

- For libraries: use the `lib` type in case you are developing a library to be consumed by others
- For end-projects: use the `end-project` type in case you developing a top-level project, such as an Web Application, a Node.js API or CLI

### 2. Setup babel within your project

The way Babel is configured depends on the the tooling you are using. Below, there are instructions for common scenarios:

#### Standard project

> If you don't use a bundler within your project, this is the setup guide you should follow

- Create `.babelrc.json` at the root of your project, replacing `<preset-type>` with the preset type you chose:

    ```json
    {
        "presets": ["@moxy/babel-preset/<preset-type>"]
    }
    ```

    ...or with options:

    ```json
    {
        "presets": [["@moxy/babel-preset/<preset-type>", { "react": true }]]
    }
    ```

- Install [`@babel/cli`](https://www.npmjs.com/package/@babel/cli) as a dev dependency because we will need it for the build script:

    ```ssh
    $ npm install @babel/cli --save-dev
    ```

- Set up your `package.json` like this:

    ```json
    "main": "lib/index.js",
    "module": "es/index.js",
    "files": [
      "lib",
      "es"
    ],
    "scripts": {
      "build:commonjs": "BABEL_ENV=commonjs babel src -d lib --delete-dir-on-start",
      "build:es": "BABEL_ENV=es babel src -d es --delete-dir-on-start",
      "build": "npm run build:commonjs && npm run build:es"
    }
    ```

    Note that the build process above will produce both CommonJS and ES builds. If you just want to produce a single build, the `package.json` may be simplified. For example, to produce a single CommonJS build:

    ```json
    "main": "lib/index.js",
    "files": [
      "lib"
    ],
    "scripts": {
      "build": "BABEL_ENV=commonjs babel src -d lib --delete-dir-on-start"
    }
    ```
- Tweak your `.gitignore` file:

   Add `lib/` and/or `es/` folder to the gitignore file so that those output folders are never committed.

- Create `src/index.js` and happy coding!

#### Webpack based project

Tweak your Webpack config JavaScript rule to include [`babel-loader`](https://www.npmjs.com/package/babel-loader) and MOXY's preset. Here's an example for a website project using React:

```js
{
    test: /\.js$/,
    use: [
        {
            loader: require.resolve('babel-loader'),
            options: {
                cacheDirectory: true,  // Improve performance
                presets: [
                    [require.resolve('@moxy/babel-preset/end-project'), {
                        targets: ['browsers'],
                        react: true,
                        modules: false,
                    }],
                ],
            },
        },
    ],
}
```

It's important that you do not exclude the `node_modules` folder so that everything goes through the `@babel/preset-env`, ensuring that all the produced code was transpiled according to the targets.

### 3. Tweak the options

Below, you may find a list containing all options you may tweak:

| Name   | Description   | Type     | Default | in `lib` | in `end-project` |
| ------ | ------------- | -------- | ------- | ------------ | ------------ |
| react | Adds support for [React](https://reactjs.org/) | boolean | false | ✅ | ✅ |
| lodash | Transform to cherry-pick Lodash modules | boolean/[Object](https://github.com/lodash/babel-plugin-lodash#usage) | true | ✅ | ✅ |
| modules | Transform ES6 module syntax to another module type | [string/boolean](https://www.npmjs.com/package/babel-preset-env#modules) | Based on `process.env.BABEL_ENV`, `commonjs` if unspecified | ✅ | ✅ |
| dynamicImport | Adds support for `import()` statements | boolean | true | ✅ | ✅ |
| loose | Enable "loose" transformations for any plugins that allow them | boolean | true | ❌ | ✅ |
| targets | The output targets, see bellow for a more detailed explanation | Array/[Object](https://babeljs.io/docs/en/next/babel-preset-env.html#targets) | ['browsers', 'node'] | ❌ | ✅ |
| env | The environment (`development`, `production` or `test`) | string | Based on `process.env.NODE_ENV` | ❌ | ✅ |
| namedDefaultExport | Use [add-module-exports](https://github.com/59naga/babel-plugin-add-module-exports) plugin to get around [babel/babel#2212](https://github.com/babel/babel/issues/2212) | boolean | true if modules is `commonjs` | ✅ | ❌ |

#### `lodash` option

Specify which modules will have the cherry-pick transformation applied.

Note that `lodash-es`, `lodash-compat` and `lodash/fp` are always added for you, regardless of having this option defined or not.

For instance, to have smaller bundles when using [recompose](https://github.com/acdlite/recompose):

```json
{
    "presets": [
        ["@moxy/babel-preset/<preset-type>", {
            "lodash": { "id": ["recompose"] }
        }],
    ],
}
```

#### `targets` option

The targets option has a very important role. By default, its value is `['browsers', 'node']` which means that the compiled code will work in both the Browser and in Node.js.

When `browsers` is specified, the compiled code will work on browsers that are supported by [Google's browser support policy](https://github.com/awkaiser/browserslist-config-google) (modern). When `node` is specified, the compiled code will work on the last LTS or higher (currently `v8.9`).

If your project has different requirements in terms of browser or node support, you may specify the [targets](https://www.npmjs.com/package/babel-preset-env#targets) yourself as an object.

#### `dynamicImport` option

Dynamic imports support are enabled by default but are dependent on the `modules` option. More specifically, the [`syntax-dynamic-import`](https://www.npmjs.com/package/@babel/plugin-syntax-dynamic-import) and [`dynamic-import-node`](https://www.npmjs.com/package/babel-plugin-transform-dynamic-import) when the `modules` option is set to `false` and `commonjs` respectively.

For other `modules` types, such as `amd`, you must find and include a plugin yourself. Also, you may disable the `dynamicImport` option by setting it to `false` in case you want to disable the feature completely or if you want to choose another plugin.

#### `env` option

The `env`'s default value respects `process.env.NODE_ENV` and falls back to `production` if none are set. When env is `production`, some plugins that perform code optimization will be enabled.

The `modules` default value is `commonjs` unless `process.env.BABEL_ENV` is set to `es`.


### 4. Be aware of the caveats

No, seriously. Read the [Caveats](#caveats) section as it contains crucial information and might require you to do a few more steps.

## Caveats

### Polyfills

#### In libraries

Shipping polyfills in libraries is, in general, a bad practice because it increases the overall file size of your top-level project due to duplication.

The [`transform-runtime`](https://www.npmjs.com/package/babel-plugin-transform-runtime) plugin attempts to solve the polyfills and duplication by transforming `Object.assign`, `Promise` and other features to their [`core-js`](https://github.com/zloirock/core-js) counter-parts. Though, this doesn't play well with [`preset-env`](https://github.com/babel/babel-preset-env/tree/1.x/) because it inlines everything, including features that are already supported by our targets. Additionally, if different versions of the runtime are installed, duplication still happens.

For this reason, you, as an author, should state in the README of your library that you expect the environment to be polyfilled with [`core-js`](https://github.com/zloirock/core-js), [`babel-polyfill`](https://babeljs.io/docs/usage/polyfill/), [`polyfill.io`](https://polyfill.io/) or similar.

#### In top-level projects

Simply include `import 'babel-polyfill';` at the top of your entry file. That statement will be replaced with the necessary polyfills based on the targets you want to support.

```js
// in:
import 'babel-polyfill';

// out:
import 'core-js/modules/es6.object.assign';
import 'core-js/modules/es6.promise';
// ...
```

### Dynamic imports

The support for dynamic imports is enabled by default, please read more on the [`dynamicImport` option](#dynamicImport).

The caveat is that [`preset-env`](https://www.npmjs.com/package/babel-preset-env) is unaware that using `import()` with Webpack relies on Promise internally. Environments which do not have builtin support for Promise, like Internet Explorer, will require both the promise and iterator polyfills be added manually. Having said that, tweak your top-level project's Webpack config like so:

```js
{
    entry: [
        'core-js/modules/es6.promise',
        'core-js/modules/es6.array.iterator',
        // Path to your entry file
    ],
};
```

### Minifying

You must use a minifier that understands ES6+ syntax because the transpiled code might contain ES6+ code.
As an example, UglifyJS v2 only understands ES5 syntax but [UglifyJS v3](https://www.npmjs.com/package/uglify-es) does support ES6+.

## Tests

```sh
$ npm test
$ npm test -- --watch # during development
```

## License

[MIT License](http://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.3.2](https://github.com/moxystudio/babel-preset/compare/v3.3.1...v3.3.2) (2021-06-05)

### [3.3.1](https://github.com/moxystudio/babel-preset/compare/v3.3.0...v3.3.1) (2020-07-31)

## [3.3.0](https://github.com/moxystudio/babel-preset/compare/v3.2.6...v3.3.0) (2020-07-31)


### Features

* change to browserslist-config-google/modern ([351da1c](https://github.com/moxystudio/babel-preset/commit/351da1cab513f25db25fb0dc42bf039962deafa1))

### [3.2.6](https://github.com/moxystudio/babel-preset/compare/v3.2.5...v3.2.6) (2020-05-28)

### [3.2.5](https://github.com/moxystudio/babel-preset/compare/v3.2.4...v3.2.5) (2020-03-10)

### [3.2.4](https://github.com/moxystudio/babel-preset/compare/v3.2.3...v3.2.4) (2020-02-01)


### Bug Fixes

* reactivate optional-chaining and nullish-coalescing-operator ([4b03a48](https://github.com/moxystudio/babel-preset/commit/4b03a4820b68e9c7ea9224b22b3c885a14c27e66))

### [3.2.3](https://github.com/moxystudio/babel-preset/compare/v3.2.2...v3.2.3) (2020-01-23)

### [3.2.2](https://github.com/moxystudio/babel-preset/compare/v3.2.1...v3.2.2) (2020-01-23)

### [3.2.1](https://github.com/moxystudio/babel-preset/compare/v3.2.0...v3.2.1) (2019-11-23)

<a name="3.2.0"></a>
# [3.2.0](https://github.com/moxystudio/babel-preset/compare/v3.1.0...v3.2.0) (2019-09-30)


### Bug Fixes

* update vulnerable/outdated dependencies ([9f44a8d](https://github.com/moxystudio/babel-preset/commit/9f44a8d))


### Features

* add nullish coalescing and optional chaining operators ([25faa38](https://github.com/moxystudio/babel-preset/commit/25faa38))



<a name="3.1.0"></a>
# [3.1.0](https://github.com/moxystudio/babel-preset/compare/v3.0.6...v3.1.0) (2019-05-28)


### Features

* enable loose option ([#44](https://github.com/moxystudio/babel-preset/issues/44)) ([9b4afe1](https://github.com/moxystudio/babel-preset/commit/9b4afe1))



<a name="3.0.6"></a>
## [3.0.6](https://github.com/moxystudio/babel-preset/compare/v3.0.5...v3.0.6) (2019-04-29)


### Bug Fixes

* add sourceType also to lib ([4b1cf95](https://github.com/moxystudio/babel-preset/commit/4b1cf95))
* set sourceType to unambiguous to properly compile deps ([b32f8c6](https://github.com/moxystudio/babel-preset/commit/b32f8c6))



<a name="3.0.5"></a>
## [3.0.5](https://github.com/moxystudio/babel-preset/compare/v3.0.4...v3.0.5) (2019-04-04)


### Bug Fixes

* add the runtime transform plugin to end project ([#42](https://github.com/moxystudio/babel-preset/issues/42)) ([8127ae7](https://github.com/moxystudio/babel-preset/commit/8127ae7))



<a name="3.0.4"></a>
## [3.0.4](https://github.com/moxystudio/babel-preset/compare/v3.0.3...v3.0.4) (2019-02-14)


### Bug Fixes

* add lib directory and update readme without preset shorthand ([#40](https://github.com/moxystudio/babel-preset/issues/40)) ([6ee6573](https://github.com/moxystudio/babel-preset/commit/6ee6573))



<a name="3.0.3"></a>
## [3.0.3](https://github.com/moxystudio/babel-preset/compare/v3.0.2...v3.0.3) (2019-02-13)



<a name="3.0.2"></a>
## [3.0.2](https://github.com/moxystudio/babel-preset/compare/v3.0.1...v3.0.2) (2019-02-13)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/moxystudio/babel-preset/compare/v3.0.0...v3.0.1) (2019-02-13)



<a name="3.0.0"></a>
# [3.0.0](https://github.com/moxystudio/babel-preset/compare/v2.3.5...v3.0.0) (2019-02-13)


### Features

* upgrade to babel 7 ([#39](https://github.com/moxystudio/babel-preset/issues/39)) ([5ad75f1](https://github.com/moxystudio/babel-preset/commit/5ad75f1)), closes [#26](https://github.com/moxystudio/babel-preset/issues/26)


### BREAKING CHANGES

* the way the preset is configured has changed



<a name="2.3.5"></a>
## [2.3.5](https://github.com/moxystudio/babel-preset/compare/v2.3.4...v2.3.5) (2018-09-11)



<a name="2.3.4"></a>
## [2.3.4](https://github.com/moxystudio/babel-preset/compare/v2.3.3...v2.3.4) (2018-09-10)


### Bug Fixes

* fix dev errors when using transform-react-inline-elements ([722526c](https://github.com/moxystudio/babel-preset/commit/722526c)), closes [#26](https://github.com/moxystudio/babel-preset/issues/26)



<a name="2.3.3"></a>
## [2.3.3](https://github.com/moxystudio/babel-preset/compare/v2.3.2...v2.3.3) (2018-03-22)


### Bug Fixes

* remove babel-plugin-transform-react-constant-elements ([#25](https://github.com/moxystudio/babel-preset/issues/25)) ([4c1ab03](https://github.com/moxystudio/babel-preset/commit/4c1ab03))



<a name="2.3.2"></a>
## [2.3.2](https://github.com/moxystudio/babel-preset/compare/v2.3.1...v2.3.2) (2018-03-11)



<a name="2.3.1"></a>
## [2.3.1](https://github.com/moxystudio/babel-preset/compare/v2.3.0...v2.3.1) (2018-03-02)



<a name="2.3.0"></a>
# [2.3.0](https://github.com/moxystudio/babel-preset/compare/v2.2.1...v2.3.0) (2018-02-28)


### Features

* add default support for async/await and generators ([96f8ce9](https://github.com/moxystudio/babel-preset/commit/96f8ce9))



<a name="2.2.1"></a>
## [2.2.1](https://github.com/moxystudio/babel-preset/compare/v2.2.0...v2.2.1) (2018-02-23)


### Bug Fixes

* fix lodash plugin being wrongly setup ([678a7c0](https://github.com/moxystudio/babel-preset/commit/678a7c0))



<a name="2.2.0"></a>
# [2.2.0](https://github.com/moxystudio/babel-preset/compare/v2.1.0...v2.2.0) (2018-02-20)


### Features

* add lodash/fp as default id on babel-plugin-lodash ([2c30efe](https://github.com/moxystudio/babel-preset/commit/2c30efe)), closes [#18](https://github.com/moxystudio/babel-preset/issues/18)



<a name="2.1.0"></a>
# [2.1.0](https://github.com/moxystudio/babel-preset/compare/v2.0.4...v2.1.0) (2018-01-24)


### Features

* add babel-plugin-lodash via options.lodash ([1104f92](https://github.com/moxystudio/babel-preset/commit/1104f92)), closes [#10](https://github.com/moxystudio/babel-preset/issues/10)



<a name="2.0.4"></a>

lib.test.js
'use strict';

const babelCore = require('@babel/core');
const serializerPath = require('jest-serializer-path');
const preset = require('./lib');

const env = process.env.NODE_ENV;

expect.addSnapshotSerializer(serializerPath);

afterEach(() => {
    process.env.NODE_ENV = env;
});

it('should have the right config with the default options', () => {
    expect(preset()).toMatchSnapshot();
});

describe('modules', () => {
    it('should set modules to false if options.modules is disabled', () => {
        expect(preset(null, { modules: false })).toMatchSnapshot();
    });

    it('should set modules to false by default if BABEL_ENV is es', () => {
        process.env.BABEL_ENV = 'es';

        try {
            expect(preset(null)).toMatchSnapshot();
        } finally {
            delete process.env.BABEL_ENV;
        }
    });
});

describe('react', () => {
    it('should enable react plugins if options.react is enabled and enable development goodies', () => {
        expect(preset(null, { react: true })).toMatchSnapshot();
    });

    it('should enable react optimizations if options.env is production', () => {
        expect(preset(null, { env: 'production', react: true })).toMatchSnapshot();
    });
});

describe('lodash', () => {
    it('should disable the lodash plugin if options.lodash is disabled', () => {
        expect(preset(null, { lodash: false })).toMatchSnapshot();
    });

    it('should pass any options to the lodash plugin', () => {
        expect(preset(null, { lodash: { id: 'recompose' } })).toMatchSnapshot();
    });
});

describe('namedDefaultExport & babel-plugin-add-module-exports', () => {
    it('should enable add-module-exports by default', () => {
        expect(preset(null)).toMatchSnapshot();
    });

    it('should enable add-module-exports if options.modules is commonjs', () => {
        expect(preset(null)).toMatchSnapshot();
        expect(preset(null, { modules: 'commonjs' })).toMatchSnapshot();
    });

    it('should not enable add-module-exports if options.modules is not commonjs', () => {
        expect(preset(null, { modules: false })).toMatchSnapshot();
    });

    it('should respect options.namedDefaultExport', () => {
        expect(preset(null, { namedDefaultExport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'commonjs', namedDefaultExport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', namedDefaultExport: true })).toMatchSnapshot();
        expect(preset(null, { namedDefaultExport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'commonjs', namedDefaultExport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', namedDefaultExport: false })).toMatchSnapshot();
    });

    it('should throw when enabling options.namedDefaultExport and options.modules is not commonjs', () => {
        expect(() => preset(null, { modules: false, namedDefaultExport: true }))
            .toThrow('The `namedDefaultExport` option can only be enabled when `modules` is commonjs');
    });
});

describe('dynamicImport', () => {
    it('should respect options.modules and options.dynamicImport', () => {
        expect(preset(null, { modules: 'commonjs', dynamicImport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', dynamicImport: true })).toMatchSnapshot();
        expect(preset(null, { modules: false, dynamicImport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'foo', dynamicImport: true })).toMatchSnapshot();

        expect(preset(null, { modules: 'commonjs', dynamicImport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', dynamicImport: false })).toMatchSnapshot();
        expect(preset(null, { modules: false, dynamicImport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'foo', dynamicImport: true })).toMatchSnapshot();
    });
});

describe('functional', () => {
    describe('class-properties', () => {
        it('should handle correctly', () => {
            expect(
                babelCore.transform(`
                    class Bork {
                        static a = 'foo';
                        static b;

                        x = 'bar';
                        y;
                    }
                `, preset()).code
            ).toMatchSnapshot();
        });
    });

    describe('optional-chaining', () => {
        it('should handle correctly', () => {
            expect(
                babelCore.transform(`
                const obj = {
                    foo: {
                      bar: {
                        baz: 42,
                      },
                    },
                };

                const baz = obj?.foo?.bar?.baz;
                `, preset()).code
            ).toMatchSnapshot();
        });
    });

    describe('nullish-coalescing-operator', () => {
        it('should handle correctly', () => {
            expect(
                babelCore.transform(`
                const foo = object.foo ?? 'bar';
                `, preset()).code
            ).toMatchSnapshot();
        });
    });
});


test/lib.spec.js
'use strict';

const babelCore = require('@babel/core');
const serializerPath = require('jest-serializer-path');
const preset = require('../lib');

const env = process.env.NODE_ENV;

expect.addSnapshotSerializer(serializerPath);

afterEach(() => {
    process.env.NODE_ENV = env;
});

it('should have the right config with the default options', () => {
    expect(preset()).toMatchSnapshot();
});

describe('modules', () => {
    it('should set modules to false if options.modules is disabled', () => {
        expect(preset(null, { modules: false })).toMatchSnapshot();
    });

    it('should set modules to false by default if BABEL_ENV is es', () => {
        process.env.BABEL_ENV = 'es';

        try {
            expect(preset(null)).toMatchSnapshot();
        } finally {
            delete process.env.BABEL_ENV;
        }
    });
});

describe('react', () => {
    it('should enable react plugins if options.react is enabled and enable development goodies', () => {
        expect(preset(null, { react: true })).toMatchSnapshot();
    });

    it('should enable react optimizations if options.env is production', () => {
        expect(preset(null, { env: 'production', react: true })).toMatchSnapshot();
    });
});

describe('lodash', () => {
    it('should disable the lodash plugin if options.lodash is disabled', () => {
        expect(preset(null, { lodash: false })).toMatchSnapshot();
    });

    it('should pass any options to the lodash plugin', () => {
        expect(preset(null, { lodash: { id: 'recompose' } })).toMatchSnapshot();
    });
});

describe('namedDefaultExport & babel-plugin-add-module-exports', () => {
    it('should enable add-module-exports by default', () => {
        expect(preset(null)).toMatchSnapshot();
    });

    it('should enable add-module-exports if options.modules is commonjs', () => {
        expect(preset(null)).toMatchSnapshot();
        expect(preset(null, { modules: 'commonjs' })).toMatchSnapshot();
    });

    it('should not enable add-module-exports if options.modules is not commonjs', () => {
        expect(preset(null, { modules: false })).toMatchSnapshot();
    });

    it('should respect options.namedDefaultExport', () => {
        expect(preset(null, { namedDefaultExport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'commonjs', namedDefaultExport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', namedDefaultExport: true })).toMatchSnapshot();
        expect(preset(null, { namedDefaultExport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'commonjs', namedDefaultExport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', namedDefaultExport: false })).toMatchSnapshot();
    });

    it('should throw when enabling options.namedDefaultExport and options.modules is not commonjs', () => {
        expect(() => preset(null, { modules: false, namedDefaultExport: true }))
            .toThrow('The `namedDefaultExport` option can only be enabled when `modules` is commonjs');
    });
});

describe('dynamicImport', () => {
    it('should respect options.modules and options.dynamicImport', () => {
        expect(preset(null, { modules: 'commonjs', dynamicImport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', dynamicImport: true })).toMatchSnapshot();
        expect(preset(null, { modules: false, dynamicImport: true })).toMatchSnapshot();
        expect(preset(null, { modules: 'foo', dynamicImport: true })).toMatchSnapshot();

        expect(preset(null, { modules: 'commonjs', dynamicImport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'cjs', dynamicImport: false })).toMatchSnapshot();
        expect(preset(null, { modules: false, dynamicImport: false })).toMatchSnapshot();
        expect(preset(null, { modules: 'foo', dynamicImport: true })).toMatchSnapshot();
    });
});

describe('functional', () => {
    describe('class-properties', () => {
        it('should handle correctly', () => {
            expect(
                babelCore.transform(`
                    class Bork {
                        static a = 'foo';
                        static b;

                        x = 'bar';
                        y;
                    }
                `, preset()).code
            ).toMatchSnapshot();
        });
    });

    describe('optional-chaining', () => {
        it('should handle correctly', () => {
            expect(
                babelCore.transform(`
                const obj = {
                    foo: {
                      bar: {
                        baz: 42,
                      },
                    },
                };

                const baz = obj?.foo?.bar?.baz;
                `, preset()).code
            ).toMatchSnapshot();
        });
    });

    describe('nullish-coalescing-operator', () => {
        it('should handle correctly', () => {
            expect(
                babelCore.transform(`
                const foo = object.foo ?? 'bar';
                `, preset()).code
            ).toMatchSnapshot();
        });
    });
});


lib/react.js
'use strict';

const addReactSupport = (config, env) => {
    // Add base support
    config.plugins.unshift(
        require.resolve('@babel/plugin-syntax-jsx'),
        [require.resolve('@babel/plugin-transform-react-jsx'), { useBuiltIns: true }],
        require.resolve('@babel/plugin-transform-react-display-name')
    );

    // Enable optimizations on production
    if (env === 'production') {
        config.plugins.push(
            [require.resolve('babel-plugin-transform-react-remove-prop-types'), { removeImport: true }]
        );
    // The following two plugins are currently necessary to make React warnings include more valuable information
    // They are included here because they are currently not enabled in babel-preset-react
    // See the below threads for more info:
    // https://github.com/babel/babel/issues/4702
    // https://github.com/babel/babel/pull/3540#issuecomment-228673661
    // https://github.com/facebookincubator/create-react-app/issues/989
    } else if (env === 'development') {
        config.plugins.push(
            // Adds component stack to warning messages
            require.resolve('@babel/plugin-transform-react-jsx-source'),
            // Adds __self attribute to JSX which React will use for some warnings
            require.resolve('@babel/plugin-transform-react-jsx-self')
        );
    }
};

module.exports = addReactSupport;


index.js
'use strict';

function addReactSupport(config, options) {
    // Add base support
    config.plugins.unshift(
        require.resolve('babel-plugin-syntax-jsx'),
        [require.resolve('babel-plugin-transform-react-jsx'), { useBuiltIns: true }],
        require.resolve('babel-plugin-transform-react-display-name')
    );

    // Enable optimizations on production
    if (options.env === 'production') {
        config.plugins.push(
            require.resolve('babel-plugin-transform-react-remove-prop-types')
        );
    // The following two plugins are currently necessary to make React warnings include more valuable information.
    // They are included here because they are currently not enabled in babel-preset-react.
    // See the below threads for more info:
    // https://github.com/babel/babel/issues/4702
    // https://github.com/babel/babel/pull/3540#issuecomment-228673661
    // https://github.com/facebookincubator/create-react-app/issues/989
    } else if (options.env === 'development') {
        config.plugins.push(
            // Adds component stack to warning messages
            require.resolve('babel-plugin-transform-react-jsx-source'),
            // Adds __self attribute to JSX which React will use for some warnings
            require.resolve('babel-plugin-transform-react-jsx-self')
        );
    }
}

// -----------------------------------------------

module.exports = (context, options) => {
    options = Object.assign({
        targets: ['browsers', 'node'], // Can be an array with `browsers` and/or `node` or an object
        modules: process.env.BABEL_ENV === 'es' ? false : 'commonjs', // Usually set to `commonjs` or `false`
        debug: false, // Enable debug mode for preset-env

        env: process.env.NODE_ENV || 'production',
        react: false,
        lodash: true,
        namedDefaultExport: null,
    }, options);

    if (options.modules !== 'commonjs' && options.namedDefaultExport) {
        throw new Error('The `namedDefaultExport` option can only be enabled when `modules` is commonjs');
    }

    // Set `namedDefaultExport` default value based on `modules`
    if (options.namedDefaultExport == null) {
        options.namedDefaultExport = options.modules === 'commonjs';
    }

    const config = {
        presets: [],
        plugins: [],
    };

    // The `preset-env` will activate the necessary features based on our targets
    // It's no longer necessary to add `es2015`, `es2016`, etc manually
    config.presets.push([require.resolve('babel-preset-env'), {
        // Replaces `import 'babel-polyfill';` with only the polyfills that are
        // actually required based on the targets
        useBuiltIns: true,
        // Produce less and more readable code (although not as faithful to the semantics)
        loose: true,
        // Print plugins & polyfills
        debug: options.debug,
        // Set modules options
        modules: options.modules,
        // Set the browser support to be the same used by Google (https://www.npmjs.com/package/browserslist-config-google)
        // Set Nodejs target to be the latest LTS
        targets: Array.isArray(options.targets) ? Object.assign({},
            options.targets.indexOf('node') !== -1 ? { node: '8.9' } : {},
            options.targets.indexOf('browsers') !== -1 ? { browsers: ['extends browserslist-config-google'] } : {}
        ) : options.targets,
    }]);

    // The two plugins above activate stage 3 features that babel hasn't added to the stage 3 preset yet
    config.plugins.push(
        // Allows class { handleClick = () => { } static propTypes = { foo: PropTypes.string } }
        require.resolve('babel-plugin-transform-class-properties'),
        // Support destructuring of objects, e.g.: { ...foo }
        [require.resolve('babel-plugin-transform-object-rest-spread'), { useBuiltIns: true }]
    );

    // Add react support
    if (options.react) {
        addReactSupport(config, options);
    }

    // Cherry-pick lodash modules for smaller bundles
    if (options.lodash) {
        // Re-include default ids plus lodash/fp
        const baseLodashOptionsIds = [
            'lodash',
            'lodash-es',
            'lodash-compat',
            'lodash/fp',
        ];

        config.plugins.push([
            require.resolve('babel-plugin-lodash'),
            Object.assign({},
                options.lodash,
                { id: baseLodashOptionsIds.concat(options.lodash.id || []) }),
        ]);
    }

    // Add `module.exports = default;`, see https://github.com/59naga/babel-plugin-add-module-exports
    if (options.namedDefaultExport) {
        config.plugins.push('babel-plugin-add-module-exports');
    }

    return config;
};
